### PR TITLE
#250 Fix autocomplete cache

### LINKED ISSUE #250: Autocomplete does not consider recently added entries
The autocomplete function for comments does not list recently added entries.

### Workaround
Restart white rabbit to get all entries.

### PR DESCRIPTION
**Pull request by kaklakariada**
Closes #250

### CODE REVIEW

File: logic/src/main/java/org/itsallcode/whiterabbit/logic/storage/MonthCache.java [hunk 1]
```diff
@@ -35,4 +39,14 @@ List<DayRecord> getLatestDays(LocalDate maxAge)
                 .filter(day -> !day.getDate().isBefore(maxAge))
                 .collect(toList());
     }
+
+    private void notifyListeners(final MonthIndex updatedMonth)
+    {
+        listeners.forEach(l -> l.cacheUpdated(updatedMonth));
```
Comment (redcatbear) [inline logic/src/main/java/org/itsallcode/whiterabbit/logic/storage/MonthCache.java:45]:
```suggestion
        listeners.forEach(listener -> listener.cacheUpdated(updatedMonth));
```

File: logic/src/test/java/org/itsallcode/whiterabbit/logic/autocomplete/AutocompleteServiceTest.java [hunk 2]
```diff
@@ -53,6 +55,27 @@ void dayCommentAutocompleter()
...
     }
 
+    @Test
+    void dayCommentAutocompleterCachesEntries()
+    {
+        simulateDays(dayRecordsWithComments(null, "", "Comment A", "Comment B"));
+        final AutocompleteEntrySupplier completer = autocompleteService.dayCommentAutocompleter();
+        completer.getEntries("comm");
+        completer.getEntries("comm");
+        verify(storageMock, times(1)).getLatestDays(any());
```
Comment (redcatbear) [inline logic/src/test/java/org/itsallcode/whiterabbit/logic/autocomplete/AutocompleteServiceTest.java:65]: Ah, I see what you did there. If you named the test something like `testSecondAutocompletedDayRecordGetsServedFromCache` the test would be easier to grasp.

**Review by redcatbear — CHANGES_REQUESTED** [review]
Some small changes requested.

File: logic/src/main/java/org/itsallcode/whiterabbit/logic/autocomplete/AutocompleteService.java [hunk 3]
```diff
@@ -30,20 +30,31 @@
...
+            this::getActivityComments);
+
+    public AutocompleteService(final CachingStorage storage, final ClockService clockService)
     {
         this.storage = storage;
         this.clockService = clockService;
+        storage.addCacheInvalidationListener(this::cacheUpdated);
+    }
+
+    private void cacheUpdated(final MonthIndex updatedMonth)
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply logic/src/main/java/org/itsallcode/whiterabbit/logic/autocomplete/AutocompleteService.java:44]: :white_check_mark:

File: api/src/main/java/org/itsallcode/whiterabbit/api/model/ProjectReportActivity.java [hunk 4]
```diff
@@ -10,16 +10,22 @@
...
 {
     /**
+     * Get the {@link Project} of this activity.
+     * 
      * @return the {@link Project} of this activity.
      */
     Project getProject();
 
     /**
+     * Get the total duration of this activity.
```
Comment (redcatbear) [inline api/src/main/java/org/itsallcode/whiterabbit/api/model/ProjectReportActivity.java:20]: Documentation seems a little bit off given the method name. But I guess it is correct.

File: jfxui/src/uiTest/java/org/itsallcode/whiterabbit/jfxui/JavaFxAppUiTest.java [hunk 5]
```diff
@@ -225,23 +226,25 @@ void selectedDayUpdatedWhenDayChanges()
...
     @Test
+    @Tag(JunitTags.FLAKY)
     void weekendsAreHighlightedAsWeekend()
     {
         final JavaFxTable<DayRecordPropertyAdapter> dayTable = app().genericDayTable();
         assertAll(
-                () -> dayTable.assertRowHasPseudoClass(0, "weekend"),
-                () -> dayTable.assertRowHasPseudoClass(1, "weekend"),
-                () -> dayTable.assertRowDoesNotHavePseudoClass(2, "weekend"));
+                () -> dayTable.assertRowHasPseudoClass(7, "weekend"),
```
Comment (redcatbear) [inline jfxui/src/uiTest/java/org/itsallcode/whiterabbit/jfxui/JavaFxAppUiTest.java:234]: This test is hard to judge if you don't document what the magic numbers are.

**Review by redcatbear — APPROVED** [review]

File: jfxui/src/uiTest/java/org/itsallcode/whiterabbit/jfxui/JavaFxAppUiTest.java (hunk 5 above)
Comment (kaklakariada) [reply jfxui/src/uiTest/java/org/itsallcode/whiterabbit/jfxui/JavaFxAppUiTest.java:234]: I will refactor the UI-Tests in a separate PR, OK?

**Pull request merged**